servicetalk-traffic-resilience-http/src/main/java/io/servicetalk/traffic/resilience/http/AbstractTrafficResilienceHttpFilter.java

```diff
@@ -249,37 +282,76 @@ private Single<StreamingHttpResponse> handleAllow(
                     }
                     return Single.succeeded(resp).shareContextOnSubscribe();
                 })
-                .liftSync(new BeforeFinallyHttpOperator(new TerminalSignalConsumer() {
-                    @Override
-                    public void onComplete() {
-                        try {
-                            if (breaker != null) {
-                                breaker.onSuccess(nanoTime() - startTimeNs, NANOSECONDS);
-                            }
-                        } finally {
-                            onSuccessTicketTerminal.accept(ticket);
-                            ticketObserver.onComplete();
-                        }
-                    }
+                .liftSync(beforeFinally(ticket, ticketObserver, breaker, startTimeNs));
+    }
 
-                    @Override
-                    public void onError(final Throwable throwable) {
-                        AbstractTrafficResilienceHttpFilter.this.onError(throwable, breaker, startTimeNs, ticket);
-                        ticketObserver.onError(throwable);
+    private Single<StreamingHttpResponse> dryRunHandleAllow(
+            final Function<HttpResponseMetaData, Duration> delayProvider, final Ticket ticket,
+            final TicketObserver ticketObserver, @Nullable final CircuitBreaker breaker, final long startTimeNs,
+            final Single<StreamingHttpResponse> responseSingle) {
+        // The map is issuing an exception that will be propagated to the downstream BeforeFinallyHttpOperator
+        // in order to invoke the appropriate callbacks to release resources.
+        // If the BeforeFinallyHttpOperator comes earlier, the Single will succeed and only downstream will
+        // see the exception, preventing callbacks to release permits; which results in the limiter eventually
+        // throttling ALL future requests.
+        // Before returning an error, we have to drain the response payload body to properly release resources
+        // and avoid leaking a connection, except for the PassthroughRequestDroppedException case.
+        return responseSingle.onErrorMap(throwable -> {
```

Comment:
This is not an equivalent replacement for a dry-run mode. If we remove the `.liftSync` that's currently applied after `flatMap`, we loose cancellation signal and `onError` that happens for payload body publisher. IIUC, everything should stay the same as under `handleAllow`, except handling inside `flatMap` for `breakerRejectionPredicate` and `capacityRejectionPredicate`.

Reply:
You're right re cancellation.

Reply:
Right, `onError` will be covered by the `liftSync` applied inside `flatMap`